### PR TITLE
feat(Voice): implement support for @discordjs/voice

### PR DESCRIPTION
**Pull request by amishshah**
**Please describe the changes this PR makes and why it should be merged:**

This PR allows for creating adapters that can be used by [@discordjs/voice](https://github.com/discordjs/voice/). This PR also removes the existing voice codebase as it will no longer be maintained.

The only voice code that will remain in discord.js will be voice states, which aren't specifically part of the Discord Voice API.

## What is @discordjs/voice?

[@discordjs/voice](https://github.com/discordjs/voice/) is a stand-alone library that implements the Discord Voice API. It aims to offer better stability, performance, and scalability than the existing voice client within Discord.js v12.

Discord.js v13 will remove support for the Voice API, and this functionality can be replaced by installing @discordjs/voice (`npm install @discordjs/voice`) as an add-on.

## Example usage:
```js
const { joinVoiceChannel } = require('@discordjs/voice');

const connection = joinVoiceChannel({
  channelId: channel.id,
  guildId: channel.guild.id,
  adapterCreator: channel.guild.voiceAdapterCreator,
});
```

## How do I update my code to use @discordjs/voice?

@discordjs/voice is nearing its release. If you want to get a head start on migrating your voice code to the library, you should take a look at this PR which shows how to create a music bot with @discordjs/voice and TypeScript - https://github.com/discordjs/voice/pull/107

**Status and versioning classification:**

- Code changes have been tested against the Discord API, or there are no code changes
- I know how to update typings and have done so, or typings don't need updating
- This PR changes the library's interface (methods or parameters added)

### CODE REVIEW

File: src/client/voice/ClientVoiceManager.js [hunk 1]
```diff
@@ -41,6 +75,27 @@ class ClientVoiceManager {
     return broadcast;
   }
 
+  /**
+   * Creates a voice adapter for a given channel, to be used by @discordjs/voice.
+   * @param {VoiceChannel} channel The voice channel to create an adapter for.
+   * @returns {DiscordGatewayAdapterCreator}
+   * @private
```
Comment (advaith1) [inline src/client/voice/ClientVoiceManager.js:82]: why is this private?

File: typings/index.d.ts [hunk 2]
```diff
@@ -1621,6 +1621,7 @@ declare module 'discord.js' {
     public type: 'voice';
     public userLimit: number;
     public join(): Promise<VoiceConnection>;
+    public createVoiceAdapter(): object;
```
Comment (advaith1) [inline typings/index.d.ts:1624]: would it be possible to import the type from `@discordjs/voice` if it is installed?

**Review by vladfrangu — CHANGES_REQUESTED** [review]

File: typings/index.d.ts (hunk 2 above)
Comment (vladfrangu) [reply typings/index.d.ts:1624]: There's no conditional type imports. We can import it directly and it'll default to `any` if not present (and you use `skipLibCheck` in your tsconfig)

File: src/client/voice/ClientVoiceManager.js (hunk 1 above)
Comment (amishshah) [reply src/client/voice/ClientVoiceManager.js:82]: It is preferred to use `channel.createVoiceAdapter` instead.

Comment (advaith1) [reply src/client/voice/ClientVoiceManager.js:82]: But the library is moving to structure methods mainly calling public methods on the Manager so users can do either way; this is similar to the Manager methods and there doesn't seem to be any disadvantage to users using this

File: typings/index.d.ts (hunk 2 above)
Comment (didinele) [reply typings/index.d.ts:1624]: > There's no conditional type imports. We can import it directly and it'll default to `any` if not present (and you use `skipLibCheck` in your tsconfig)

This is already the case with `@types/ws` being a dev dep. I see no issue doing the same with `@discordjs/voice`?

File: src/client/voice/ClientVoiceManager.js (hunk 1 above)
Comment (amishshah) [reply src/client/voice/ClientVoiceManager.js]: The code has been refactored to remove 2 declarations of this

**Comment by amishshah** [comment]
I'm converting this back to a draft as I want to fix https://github.com/discordjs/voice/issues/78 before adapters are added directly to Discord.js.

File: src/client/voice/ClientVoiceManager.js [hunk 3]
```diff
@@ -29,6 +35,36 @@ class ClientVoiceManager {
...
+        const guild = client.guilds.cache.get(guildID);
+        if (guild?.shardID === id) {
+          adapter.destroy();
+          this.adapters.delete(guildID);
+        }
+      }
+    };
+
+    client.on('shardDisconnect', (event, id) => onShardDisconnect(id));
+    client.on('shardReconnecting', id => onShardDisconnect(id));
```
Comment (vladfrangu) [inline src/client/voice/ClientVoiceManager.js:67]: So... on reconnecting we destroy it entirely...will it reconnect / recreate after it successfully reconnects?

Comment (amishshah) [reply src/client/voice/ClientVoiceManager.js:67]: That's up to the developer (if they notice that a particular shard has reconnected, they can respawn any disconnected voice connections if they want to)

**Review by vladfrangu — APPROVED** [review]

**Comment by amishshah** [comment]
Converting to a draft, need to update the adapter as well as deprecate current Discord.js voice code.

File: src/structures/BaseGuildVoiceChannel.js [hunk 4]
```diff
@@ -66,30 +66,6 @@ class BaseGuildVoiceChannel extends GuildChannel {
...
-    return this.client.voice.joinChannel(this);
-  }
-
-  /**
-   * Leaves this voice-based channel.
-   * @example
-   * // Leave a voice-based channel
-   * channel.leave();
-   */
-  leave() {
```
Comment (kyranet) [inline src/structures/BaseGuildVoiceChannel.js:88]: Without join and leave, how are developers supposed to join/leave voice channels?

Comment (amishshah) [reply src/structures/BaseGuildVoiceChannel.js:88]: That is done completely externally now, e.g. with @discordjs/voice

File: typings/index.d.ts [hunk 5]
```diff
@@ -866,6 +855,7 @@ declare module 'discord.js' {
     public setWidget(widget: GuildWidgetData, reason?: string): Promise<Guild>;
     public splashURL(options?: ImageURLOptions): string | null;
     public toJSON(): unknown;
+    public readonly voiceAdapterCreator: DiscordGatewayAdapterCreator;
```
Comment (SpaceEEC) [inline typings/index.d.ts:858]: This should be above the methods. 👀

File: src/client/voice/ClientVoiceManager.js [hunk 6]
```diff
@@ -19,97 +16,29 @@ class ClientVoiceManager {
...
-     * @type {Collection<Snowflake, VoiceConnection>}
-     */
-    this.connections = new Collection();
-
-    /**
-     * Active voice broadcasts that have been created
-     * @type {VoiceBroadcast[]}
+     * Maps guild IDs to voice adapters created for use with @discordjs/voice.
+     * @type {Map<Snowflake, Object>}
+     * @readonly
```
Comment (SpaceEEC) [inline src/client/voice/ClientVoiceManager.js:21]: We only `@readonly` properties if they are actually read-only.
```suggestion
```

File: src/client/voice/ClientVoiceManager.js [hunk 7]
```diff
@@ -19,97 +16,29 @@ class ClientVoiceManager {
...
      */
-    this.broadcasts = [];
-  }
+    this.adapters = new Map();
```
Comment (SpaceEEC) [inline src/client/voice/ClientVoiceManager.js:23]: This does not seem to be in the typings 👀

**Review by SpaceEEC — CHANGES_REQUESTED** [review]

**Review by kyranet — APPROVED** [review]
~~I'm not sure if the lack of `@discordjs/voice` in devDeps is intentional, but otherwise~~ LGTM.

Edit: I just noticed the package.json change 😅

**Review by vladfrangu — APPROVED** [review]

**Review by iCrawl — APPROVED** [review]

**Review by SpaceEEC — APPROVED** [review]

**Pull request merged**